File: backend/schema/movimentacao_schema.py
from pydantic import BaseModel, Field, conint
from typing import Literal
from datetime import datetime  

StrictPositiveInt = conint(strict=True, gt=0)

class MovimentacaoBase(BaseModel):
    tipo: Literal["entrada", "saida"]
    quantidade: StrictPositiveInt = Field(..., description="Deve ser > 0")
    produto_id: StrictPositiveInt = Field(..., description="ID do produto válido")

class MovimentacaoCreate(MovimentacaoBase):
    pass  

class MovimentacaoRead(MovimentacaoBase):
    id: int
    data: datetime 
    produto_nome: str | None = None

    model_config = {
        "from_attributes": True
    }

File: frontend/src/schema/movimentacao_schema.ts
import { z } from 'zod';

export const tipoMovimentacaoEnum = z.enum(['entrada', 'saida'], {
  required_error: 'O tipo da movimentação é obrigatório.',
});

export const movimentacaoFormSchema = z.object({
  produto_id: z
    .union([z.string().regex(/^\d+$/, 'Produto inválido'), z.number()])
    .transform((val) => typeof val === 'string' ? parseInt(val, 10) : val)
    .refine((val) => val > 0, { message: 'Produto inválido.' }),

  tipo: tipoMovimentacaoEnum,

  quantidade: z
    .number({ invalid_type_error: 'Informe um número válido.' })
    .int('A quantidade deve ser um número inteiro.')
    .positive('A quantidade deve ser maior que zero.'),
});

export type MovimentacaoFormData = z.infer<typeof movimentacaoFormSchema>;

export interface Movimentacao extends MovimentacaoFormData {
  id: number;
  data: string; // ISO date
  produto_nome?: string;
}

export type MovimentacaoFormState = {
  produto_id: string;
  tipo: 'entrada' | 'saida';
  quantidade: number;
};
